File: lib/models/marketplace/listing.dart
/// Savdo (marketplace) modellari — buyum e'lonlari.
///
/// Backend: /api/v1/marketplace/*
///
/// DIQQAT: bu modelda sotuvchining TELEFON RAQAMI yo'q va bo'lmaydi
/// ham — backend uni umuman yubormaydi. Aloqa faqat ilova ichida.
library;

enum ListingStatus { active, sold, expired, hidden, unknown }

ListingStatus listingStatusFrom(String? raw) {
  switch (raw) {
    case 'active':
      return ListingStatus.active;
    case 'sold':
      return ListingStatus.sold;
    case 'expired':
      return ListingStatus.expired;
    case 'hidden':
      return ListingStatus.hidden;
    default:
      return ListingStatus.unknown;
  }
}

String listingStatusLabel(ListingStatus s) {
  switch (s) {
    case ListingStatus.active:
      return 'Faol';
    case ListingStatus.sold:
      return 'Sotildi';
    case ListingStatus.expired:
      return 'Muddati tugagan';
    case ListingStatus.hidden:
      return 'Yashirilgan';
    case ListingStatus.unknown:
      return '—';
  }
}

/// Buyum holati — xaridor uchun eng muhim ma'lumotlardan biri.
String conditionLabel(String? raw) {
  switch (raw) {
    case 'new':
      return 'Yangi';
    case 'like_new':
      return 'Ideal';
    case 'good':
      return 'Yaxshi';
    case 'used':
      return 'Ishlatilgan';
    case 'parts':
      return 'Ehtiyot qismga';
    default:
      return '—';
  }
}

class Listing {
  final int id;
  final int userId;
  final String categoryKey;
  final String title;
  final String description;

  /// Sotuvchi kiritgan narx va valyuta (UZS yoki USD).
  final double? price;
  final String currency;

  /// Xaridorga KO'RSATILADIGAN narx — doim so'mda (backend hisoblaydi).
  final double? priceUzs;
  final bool isNegotiable;

  final String? condition;
  final Map<String, dynamic> attributes;
  final String address;
  final double? lat;
  final double? lng;
  final ListingStatus status;
  final int views;
  final List<String> photos;
  final DateTime? expiresAt;
  final DateTime? createdAt;

  /// Xaridordan qancha uzoqda (km). Koordinata bo'lgandagina keladi.
  final double? distanceKm;

  const Listing({
    required this.id,
    required this.userId,
    required this.categoryKey,
    required this.title,
    required this.description,
    required this.price,
    required this.currency,
    required this.priceUzs,
    required this.isNegotiable,
    required this.condition,
    required this.attributes,
    required this.address,
    required this.lat,
    required this.lng,
    required this.status,
    required this.views,
    required this.photos,
    required this.expiresAt,
    required this.createdAt,
    required this.distanceKm,
  });

  factory Listing.fromJson(Map<String, dynamic> json) {
    return Listing(
      id: (json['id'] as num?)?.toInt() ?? 0,
      userId: (json['user_id'] as num?)?.toInt() ?? 0,
      categoryKey: json['category_key'] as String? ?? 'boshqa',
      title: json['title'] as String? ?? '',
      description: json['description'] as String? ?? '',
      price: (json['price'] as num?)?.toDouble(),
      currency: json['currency'] as String? ?? 'UZS',
      priceUzs: (json['price_uzs'] as num?)?.toDouble(),
      isNegotiable: json['is_negotiable'] == true,
      condition: json['condition'] as String?,
      attributes: Map<String, dynamic>.from(
        (json['attributes'] as Map?) ?? const {},
      ),
      address: json['address'] as String? ?? '',
      lat: (json['lat'] as num?)?.toDouble(),
      lng: (json['lng'] as num?)?.toDouble(),
      status: listingStatusFrom(json['status'] as String?),
      views: (json['views'] as num?)?.toInt() ?? 0,
      photos: ((json['photos'] as List?) ?? const [])
          .map((e) => e.toString())
          .toList(),
      expiresAt: _date(json['expires_at']),
      createdAt: _date(json['created_at']),
      distanceKm: (json['distance_km'] as num?)?.toDouble(),
    );
  }

  static DateTime? _date(dynamic raw) {
    if (raw == null) return null;
    return DateTime.tryParse(raw.toString());
  }

  /// Kartada ko'rinadigan asosiy rasm (birinchisi).
  String? get mainPhoto => photos.isEmpty ? null : photos.first;

  /// Narx matni — DOIM so'mda. Dollarli e'lon bo'lsa asli qavsda.
  String get priceText {
    if (isNegotiable || priceUzs == null) return 'Kelishamiz';
    final asosiy = '${_thousands(priceUzs!)} so\'m';
    if (currency == 'USD' && price != null) {
      return '$asosiy (${_thousands(price!)} \$)';
    }
    return asosiy;
  }

  String get conditionText => conditionLabel(condition);

  String get distanceText =>
      distanceKm == null ? '' : '${distanceKm!.toStringAsFixed(1)} km';

  /// Muddat tugashiga qolgan kun. Tugagan bo'lsa 0.
  int get daysLeft {
    if (expiresAt == null) return 0;
    final qolgan = expiresAt!.difference(DateTime.now()).inDays;
    return qolgan < 0 ? 0 : qolgan;
  }

  bool get isExpired =>
      status == ListingStatus.expired ||
      (expiresAt != null && expiresAt!.isBefore(DateTime.now()));

  static String _thousands(double value) {
    final butun = value.round().toString();
    final buf = StringBuffer();
    for (var i = 0; i < butun.length; i++) {
      if (i > 0 && (butun.length - i) % 3 == 0) buf.write(' ');
      buf.write(butun[i]);
    }
    return buf.toString();
  }
}

File: backend/app/models/marketplace/listing.py
"""E'lon (mahsulot) modeli — OLX uslubidagi savdo.

Nega `jobs` dan alohida: `job_posts` — bu ISH e'loni (mijoz usta
qidiradi). Bu yerda esa BUYUM sotiladi. Ikkalasining maydonlari,
qoidalari va oqimi boshqacha, aralashtirilsa ikkalasi ham chalkashadi.
"""
from datetime import datetime, timezone
import enum

from sqlalchemy import (
    JSON, DateTime, Enum, Float, ForeignKey, Integer, String, Text, func,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class ListingStatus(str, enum.Enum):
    """E'lon holati."""

    active = "active"        # Faol, qidiruvda ko'rinadi
    sold = "sold"            # Sotildi
    expired = "expired"      # Muddati tugadi (uzaytirish mumkin)
    hidden = "hidden"        # Egasi vaqtincha yashirdi


class ListingCondition(str, enum.Enum):
    """Buyum holati — xaridor uchun eng muhim ma'lumotlardan biri."""

    new = "new"              # Yangi, ishlatilmagan
    like_new = "like_new"    # Ideal holatda
    good = "good"            # Yaxshi
    used = "used"            # Ishlatilgan
    parts = "parts"          # Ehtiyot qismga


class Listing(Base):
    """Sotuvga qo'yilgan buyum."""

    __tablename__ = "listings"

    id: Mapped[int] = mapped_column(Integer, primary_key=True,
                                    autoincrement=True)
    user_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("users.id"), index=True
    )

    # Toifa — `marketplace/fields.py` dagi kalit (telefon, avto...).
    # Category jadvaliga BOG'LANMAYDI: u xizmatlar uchun.
    category_key: Mapped[str] = mapped_column(String(50), index=True)

    title: Mapped[str] = mapped_column(String(200))
    description: Mapped[str] = mapped_column(Text)

    # Narx. `price = None` — "Kelishamiz" (foydalanuvchi so'ragan).
    price: Mapped[float | None] = mapped_column(Float, nullable=True)
    # Sotuvchi qaysi valyutada kiritgani. Xaridorga DOIM so'mda
    # ko'rsatiladi (currency.py konvertatsiya qiladi).
    currency: Mapped[str] = mapped_column(String(3), default="UZS")
    is_negotiable: Mapped[bool] = mapped_column(default=False)

    condition: Mapped[ListingCondition] = mapped_column(
        Enum(ListingCondition, native_enum=False, length=20),
        default=ListingCondition.used,
    )

    # Toifaga xos maydonlar: xotira, yil, probeg, o'lcham...
    # Har toifa uchun ro'yxat `fields.py` da.
    attributes: Mapped[dict | None] = mapped_column(JSON, nullable=True)

    address: Mapped[str] = mapped_column(String(500))
    # Koordinata ixtiyoriy: bo'lsa "yaqindagilar" saralanadi.
    lat: Mapped[float | None] = mapped_column(Float, nullable=True)
    lng: Mapped[float | None] = mapped_column(Float, nullable=True)

    status: Mapped[ListingStatus] = mapped_column(
        Enum(ListingStatus, native_enum=False, length=20),
        default=ListingStatus.active,
        index=True,
    )
    views: Mapped[int] = mapped_column(Integer, default=0)

    # Muddat: oddiy foydalanuvchiga qisqa, premiumga uzoq.
    # Tugagach e'lon o'chmaydi — "Mening e'lonlarim" da uzaytiriladi.
    expires_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True, index=True
    )
    sold_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )

    photos = relationship(
        "ListingPhoto",
        back_populates="listing",
        lazy="selectin",
        cascade="all, delete-orphan",
        order_by="ListingPhoto.sort_order",
    )

    def is_open(self, now: datetime | None = None) -> bool:
        """E'lon hozir xaridorlarga ko'rinadimi."""
        if self.status != ListingStatus.active:
            return False
        if self.expires_at is None:
            return True
        hozir = now or datetime.now(timezone.utc)
        muddat = self.expires_at
        # Bazadan tz'siz kelishi mumkin — solishtirishda TypeError
        # bo'lmasligi uchun UTC deb qaraymiz.
        if muddat.tzinfo is None:
            muddat = muddat.replace(tzinfo=timezone.utc)
        return muddat > hozir

    def to_dict(self, *, distance_km: float | None = None,
                price_uzs: float | None = None) -> dict:
        """Mijozga yuboriladigan ko'rinish.

        DIQQAT: sotuvchining TELEFON RAQAMI bu yerda YO'Q va
        qo'shilmasligi kerak. Aloqa faqat ilova ichida (chat/qo'ng'iroq).
        Raqam tarqalsa firibgarlik oshadi va biz oqimdan chiqib qolamiz.
        """
        return {
            "id": self.id,
            "user_id": self.user_id,
            "category_key": self.category_key,
            "title": self.title,
            "description": self.description,
            "price": self.price,
            "currency": self.currency,
            # Xaridor doim so'mda ko'radi (chalg'imasin).
            "price_uzs": price_uzs,
            "is_negotiable": self.is_negotiable,
            "condition": self.condition.value if self.condition else None,
            "attributes": self.attributes or {},
            "address": self.address,
            "lat": self.lat,
            "lng": self.lng,
            "status": self.status.value if self.status else None,
            "views": self.views,
            "photos": [p.url for p in (self.photos or [])],
            "expires_at": (self.expires_at.isoformat()
                           if self.expires_at else None),
            "created_at": (self.created_at.isoformat()
                           if self.created_at else None),
            "distance_km": distance_km,
        }
